Std/Data/BitVec/Bitblast.lean

```diff
@@ -0,0 +1,180 @@
+/-
+Copyright (c) 2023 by the authors listed in the file AUTHORS and their
+institutional affiliations. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Harun Khan, Abdalrhman M Mohamed, Joe Hendrix
+-/
+import Std.Data.BitVec.Folds
+
+/-!
+# Bitblasting of bitvectors
+
+We prove the equivalence of a bitblasting operations and their `BitVec` implementation.
+
+## Main results
+* `BV_add`: Bitblasted addition `bitadd` is equivalent to `BitVec.add`.
+
+## Future work
+All other operations are to be PR'ed later and are already proved in
+https://github.com/mhk119/lean-smt/blob/bitvec/Smt/Data/BitVec.lean.
+-/
+
+namespace Std
+
+open Nat Bool
+
+/-! ### Preliminaries -/
+
+namespace BitVec
+
+private
+theorem testBit_limit {x i : Nat} (x_lt_succ : x < 2^(i+1)) : testBit x i = decide (x ≥ 2^i) := by
+  cases xi : testBit x i with
+  | true =>
+    simp [testBit_implies_ge xi]
+  | false =>
+    simp
+    cases Nat.lt_or_ge x (2^i) with
+    | inl x_lt =>
+      exact x_lt
+    | inr x_ge =>
+      have ⟨j, ⟨j_ge, jp⟩⟩  := ge_two_pow_implies_high_bit_true x_ge
+      cases Nat.lt_or_eq_of_le j_ge with
+      | inr x_eq =>
+        simp [x_eq, jp] at xi
+      | inl x_lt =>
+        exfalso
+        apply Nat.lt_irrefl
+        exact
+            calc x < 2^(i+1) := x_lt_succ
+                 _ ≤ 2 ^ j := Nat.pow_le_pow_of_le_right Nat.zero_lt_two x_lt
+                 _ ≤ x := testBit_implies_ge jp
+
+private
+theorem mod_two_pow_succ (x i : Nat) :
+  x % 2^(i+1) = 2^i*(x.testBit i).toNat + x % (2 ^ i):= by
+  apply Nat.eq_of_testBit_eq
+  intro j
+  simp only [
+        Nat.mul_add_lt_is_or,
+        testBit_or,
+        testBit_mod_two_pow,
+        testBit_shiftLeft,
+        Nat.testBit_bool_to_nat,
+        Nat.sub_eq_zero_iff_le,
+        Nat.mod_lt,
+        Nat.pow_two_pos,
+        testBit_mul_pow_two
+        ]
+  match Nat.lt_trichotomy i j with
+  | Or.inl i_lt_j =>
+    have i_le_j : i ≤ j := Nat.le_of_lt i_lt_j
+    have not_j_le_i : ¬(j ≤ i) := Nat.not_le_of_lt i_lt_j
+    have not_j_lt_i : ¬(j < i) := Nat.not_lt_of_le i_le_j
+    have not_j_lt_i_succ : ¬(j < i + 1) :=
+          Nat.not_le_of_lt (Nat.succ_lt_succ i_lt_j)
+    simp [i_le_j, not_j_le_i, not_j_lt_i, not_j_lt_i_succ]
+
+  | Or.inr (Or.inl i_eq_j) =>
+    simp [i_eq_j]
+  | Or.inr (Or.inr j_lt_i) =>
+    have j_le_i : j ≤ i := Nat.le_of_lt j_lt_i
+    have j_le_i_succ : j < i + 1 := Nat.succ_le_succ j_le_i
+    have not_j_ge_i : ¬(j ≥ i) := Nat.not_le_of_lt j_lt_i
+    simp [j_lt_i, j_le_i, not_j_ge_i, j_le_i_succ]
+
+private
+theorem mod_two_pow_lt (x i : Nat) : x % 2 ^ i < 2^i := Nat.mod_lt _ (Nat.pow_two_pos _)
```

Comment:
The usual Std practice is to put modifiers (e.g. `private`, `@[simp]`) on the same line, unless it results in really gross linewrapping.

Again, I don't particular mind if we relax this one.